code/datums/components/light_eater.dm

```diff
@@ -0,0 +1,198 @@
+		return COMPONENT_INCOMPATIBLE
+
+	if(length(_eaten))
+		eaten_lights = list()
+		var/list/cached_eaten_lights = eaten_lights
+		for(var/food in _eaten)
+			RegisterSignal(food, COMSIG_ATOM_UPDATE_LIGHT, .proc/block_light_update)
+			RegisterSignal(food, COMSIG_PARENT_QDELETING, .proc/deref_eaten_light)
+			RegisterSignal(food, COMSIG_PARENT_EXAMINE, .proc/on_eaten_light_examine)
+			cached_eaten_lights += food
+	return ..()
+
+/datum/component/light_eater/Destroy(force, silent)
+	var/list/signals_to_unregister = list(
+		COMSIG_ATOM_UPDATE_LIGHT,
+		COMSIG_PARENT_QDELETING,
+	)
+	for(var/light in eaten_lights)
+		UnregisterSignal(light, signals_to_unregister)
+	eaten_lights = null
+	return ..()
+
+/datum/component/light_eater/RegisterWithParent()
+	. = ..()
+	if(isatom(parent))
+		if(ismovable(parent))
+			RegisterSignal(parent, COMSIG_MOVABLE_IMPACT, .proc/on_throw_impact)
+			if(isitem(parent))
+				RegisterSignal(parent, COMSIG_ITEM_AFTERATTACK, .proc/on_afterattack)
+				RegisterSignal(parent, COMSIG_ITEM_HIT_REACT, .proc/on_hit_reaction)
+			else if(isprojectile(parent))
+				RegisterSignal(parent, COMSIG_PROJECTILE_ON_HIT, .proc/on_projectile_hit)
+	else if(istype(parent, /datum/reagent))
+		RegisterSignal(parent, COMSIG_REAGENT_EXPOSE_ATOM, .proc/on_expose_atom)
+
+/datum/component/light_eater/UnregisterFromParent()
+	. = ..()
+	if(isatom(parent))
+		if(ismovable(parent))
+			UnregisterSignal(parent, COMSIG_MOVABLE_IMPACT)
+			if(isitem(parent))
+				UnregisterSignal(parent, list(
+					COMSIG_ITEM_AFTERATTACK,
+					COMSIG_ITEM_HIT_REACT,
+				))
+			else if(isprojectile(parent))
+				UnregisterSignal(parent, COMSIG_PROJECTILE_ON_HIT)
+	else if(istype(parent, /datum/reagent))
+		UnregisterSignal(parent, COMSIG_REAGENT_EXPOSE_ATOM)
+
+/datum/component/light_eater/InheritComponent(datum/component/C, i_am_original, list/_eaten)
+	. = ..()
+	if(length(_eaten))
+		LAZYINITLIST(eaten_lights)
+		var/list/cached_eaten_lights = eaten_lights
+		for(var/food in _eaten)
+			RegisterSignal(food, COMSIG_ATOM_UPDATE_LIGHT, .proc/block_light_update)
+			RegisterSignal(food, COMSIG_PARENT_QDELETING, .proc/deref_eaten_light)
+			RegisterSignal(food, COMSIG_PARENT_EXAMINE, .proc/on_eaten_light_examine)
+			cached_eaten_lights += food
+
+/// Handles queuing lights to eat
+/datum/component/light_eater/proc/eat_lights(atom/food, datum/eater)
+	var/list/buffet = light_eater_table_buffet(food, eater)
+	for(var/morsel in buffet)
+		devour(morsel, eater)
+
+/// Handles eating lights
+/datum/component/light_eater/proc/devour(atom/morsel, datum/eater)
+	if(!light_eater_devour(morsel, eater))
+		return FALSE
+
+	morsel.visible_message(
+		"<span class='danger'>Something dark and hungry swarms out of \the [eater] and over \the [morsel]!</span>",
+		"<span class='userdanger'>Something dark and hungry swarms out of \the [eater] and burrows into you!</span>",
+		"<span class='danger'>You can feel a dark hum gnaw at your sight.</span>"
+	)
```

Comment:
This might get spammy. You might want to compile this as a list and return it for feedback, perhaps, so a single visible message is sent.

Reply:
Done